frontend/app/src/main/java/com/example/speechbuddy/viewmodel/SignupViewModel.kt

```diff
@@ -0,0 +1,152 @@
+import com.example.speechbuddy.ui.models.SignupError
+import com.example.speechbuddy.ui.models.SignupErrorType
+import com.example.speechbuddy.ui.models.SignupUiState
+import com.example.speechbuddy.utils.Constants
+import com.example.speechbuddy.utils.Resource
+import com.example.speechbuddy.utils.isValidPassword
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class SignupViewModel @Inject internal constructor(
+    private val repository: AuthRepository
+) : ViewModel() {
+
+    private val _uiState = MutableStateFlow(SignupUiState())
+    val uiState: StateFlow<SignupUiState> = _uiState.asStateFlow()
+
+    var emailInput by mutableStateOf("")
+        private set
+
+    var nicknameInput by mutableStateOf("")
+        private set
+
+    var passwordInput by mutableStateOf("")
+        private set
+
+    var passwordCheckInput by mutableStateOf("")
+        private set
+
+    private val _signupResult = MutableLiveData<Resource<Void>>()
+    val signupResult: LiveData<Resource<Void>> = _signupResult
+
+    fun setNickname(input: String) {
+        nicknameInput = input
+        if (_uiState.value.error?.type == SignupErrorType.NICKNAME) validateNickname()
+    }
+
+    fun setPassword(input: String) {
+        passwordInput = input
+        if (_uiState.value.error?.type == SignupErrorType.PASSWORD) validatePassword()
+    }
+
+    fun setPasswordCheck(input: String) {
+        passwordCheckInput = input
+        if (_uiState.value.error?.type == SignupErrorType.PASSWORDCHECK) validatePasswordCheck()
+    }
+
+    private fun validateNickname() {
+        if (nicknameInput.isNotEmpty() && nicknameInput.length <= Constants.MAXIMUM_NICKNAME_LENGTH) {
+            _uiState.update { currentSate ->
+                currentSate.copy(
+                    isValidNickname = true,
+                    error = null
+                )
+            }
+        }
+    }
+
+    // Check password length
+    private fun validatePassword() {
+        if (isValidPassword(passwordInput)) {
+            _uiState.update { currentState ->
+                currentState.copy(
+                    isValidPassword = true,
+                    error = null
+                )
+            }
+        }
+    }
+
+    // Check password equality
+    private fun validatePasswordCheck() {
+        if (passwordInput == passwordCheckInput) {
+            _uiState.update { currentState ->
+                currentState.copy(
+                    isValidEmail = true,
+                    error = null
+                )
+            }
+        }
+    }
+
+    fun clearInputs() {
+        nicknameInput = ""
+        passwordInput = ""
+        passwordCheckInput = ""
+    }
+
+    fun signUp() {
```

Comment:
```suggestion
    fun signup(emailInput: String) {
```

1. login, signup 등은 한 단어로 통일해서 쓰는 게 가독성이 좋은 것 같습니다
2. (사용자가 이메일 인증 단계에서 입력한) emailInput이 navController로 ui level에서 전달되기 때문에 이를 viewmodel로 내려주기 위해서는 SignupScreen이 사용할 `signup()` 메소드에 파라미터를 두면 될 것 같습니다